### PR TITLE
`src/cdef.rs`: Deduplicate w/ generics 

### PR DESCRIPTION
**Pull request by thedataking**

### CODE REVIEW

**Review by kkysen — COMMENTED** [review]
By the way, how did you get git to treat the deduplicated file as a moved and edited version of the `_tmpl_16.rs` file?  It's much easier to review that way and spot differences that could be causing the arm CI failures.

File: src/cdef.rs [hunk 1]
```diff
@@ -549,3 +556,703 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    damping: c_int,
+    edges: CdefEdgeFlags,
+    _bitdepth_max: c_int,
+) {
+    use crate::src::align::Align16;
+
+    let mut tmp_buf = Align16([0; 200]);
+    let tmp = tmp_buf.0.as_mut_ptr().offset(2 * 16).offset(8);
+    dav1d_cdef_padding8_8bpc_neon(tmp, dst, stride, left, top, bottom, 8, edges);
+    dav1d_cdef_filter8_8bpc_neon(
```
Comment (kkysen) [inline src/cdef.rs:1146]: You're calling the bitdepth-specific versions here; this `fn` needs to be generic and dispatch on `BD::BPC`.

**Review by kkysen — CHANGES_REQUESTED** [review]
The `fn *_neon_erased`s aren't bitdepth generic and aren't dispatching to the correct bitdepth `fn`s (just one of them).

**Comment by thedataking** [comment]
> By the way, how did you get git to treat the deduplicated file as a moved and edited version of the _tmpl_16.rs file?

No idea. I just used `git rm ...` 🤷‍♂️  

> The fn *_neon_eraseds aren't bitdepth generic and aren't dispatching to the correct bitdepth fns (just one of them).

yeah, should have waited for the CI to finish before pinging you; thanks!

File: src/cdef.rs [hunk 2]
```diff
@@ -1,11 +1,27 @@
+use crate::include::common::attributes::clz;
+use crate::include::common::bitdepth::AsPrimitive;
+use crate::include::common::bitdepth::BitDepth;
 use crate::include::common::bitdepth::DynPixel;
 use crate::include::common::bitdepth::LeftPixelRow2px;
 use crate::include::common::intops::apply_sign;
+use crate::include::common::intops::iclip;
+use crate::include::common::intops::ulog2;
+use crate::src::tables::dav1d_cdef_directions;
+
```
Comment (kkysen) [inline src/cdef.rs:10]:
```suggestion
```

File: src/cdef.rs [hunk 3]
```diff
@@ -549,3 +559,750 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    let mut tmp_buf: [i16; 144] = [0; 144];
+    let mut tmp: *mut i16 = tmp_buf
+        .as_mut_ptr()
+        .offset((2 * tmp_stride) as isize)
+        .offset(2);
+    padding::<BD>(
+        tmp, tmp_stride, dst, dst_stride, left, top, bottom, w, h, edges,
+    );
+    if pri_strength != 0 {
+        let bitdepth_min_8 = 32 - clz(bd.bitdepth_max().as_::<c_uint>()) - 8;
```
Comment (kkysen) [inline src/cdef.rs:692]:
```suggestion
        let bitdepth_min_8 = bd.bitdepth().as_::<c_int>() - 8;
```

File: src/cdef.rs [hunk 4]
```diff
@@ -549,3 +559,750 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    cdef_find_dir_rust(img.cast(), stride, var, BD::from_c(bitdepth_max))
+}
+
+unsafe fn cdef_find_dir_rust<BD: BitDepth>(
+    mut img: *const BD::Pixel,
+    stride: ptrdiff_t,
+    var: *mut c_uint,
+    bd: BD,
+) -> c_int {
+    let bitdepth_min_8 = 32 - clz(bd.bitdepth_max().as_::<c_uint>()) - 8;
```
Comment (kkysen) [inline src/cdef.rs:923]:
```suggestion
    let bitdepth_min_8 = bd.bitdepth().as_::<c_int>() - 8;
```

File: src/cdef.rs [hunk 5]
```diff
@@ -549,3 +559,750 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    top: *const DynPixel,
+    bottom: *const DynPixel,
+    pri_strength: c_int,
+    sec_strength: c_int,
+    dir: c_int,
+    damping: c_int,
+    edges: CdefEdgeFlags,
+    bitdepth_max: c_int,
+) {
+    cdef_filter_block_c::<BD>(
```
Comment (kkysen) [inline src/cdef.rs:831]:
```suggestion
    cdef_filter_block_c(
```

File: src/cdef.rs [hunk 6]
```diff
@@ -1,11 +1,27 @@
...
 use std::ffi::c_int;
 use std::ffi::c_uint;
 
+#[cfg(feature = "asm")]
+use cfg_if::cfg_if;
+
+#[cfg(feature = "asm")]
+use crate::src::cpu::{rav1d_get_cpu_flags, CpuFlags};
+
+#[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64"),))]
```
Comment (kkysen) [inline src/cdef.rs:22]:
```suggestion
#[cfg(feature = "asm")]
```

See below for the other import removed.

File: src/cdef.rs [hunk 7]
```diff
@@ -549,3 +559,750 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    }
+    *var = best_cost.wrapping_sub(cost[(best_dir ^ 4 as c_int) as usize]) >> 10;
+    return best_dir;
+}
+
+#[inline(always)]
+#[cfg(all(feature = "asm", any(target_arch = "x86", target_arch = "x86_64"),))]
+unsafe fn cdef_dsp_init_x86<BD: BitDepth>(c: *mut Rav1dCdefDSPContext) {
+    use crate::include::common::bitdepth::BPC;
+
```
Comment (kkysen) [inline src/cdef.rs:1033]:
```suggestion
```

File: src/cdef.rs [hunk 8]
```diff
@@ -549,3 +559,750 @@ pub unsafe fn fill(mut tmp: *mut i16, stride: ptrdiff_t, w: c_int, h: c_int) {
...
+    let flags = rav1d_get_cpu_flags();
+
+    if !flags.contains(CpuFlags::SSE2) {
+        return;
+    }
+
+    match BD::BPC {
+        BPC::BPC8 => {
+            (*c).fb[0] = dav1d_cdef_filter_8x8_8bpc_sse2;
```
Comment (kkysen) [inline src/cdef.rs:1042]:
```suggestion
            if !flags.contains(CpuFlags::SSSE2) {
                return;
            }

            (*c).fb[0] = dav1d_cdef_filter_8x8_8bpc_sse2;
```

Comment (kkysen) [reply src/cdef.rs:1042]: By the way, this logic is harder to review/map back to the C when the two bitdepth branches are fully separated, i.e. one `match` vs. a `match` for each `fn` ptr (like how `bd_fn!` does it).  It's fine for now, though.

**Review by kkysen — CHANGES_REQUESTED** [review]
Just a couple of little changes left.

**Review by kkysen — APPROVED** [review]

**Pull request merged**